### PR TITLE
sync apidef templates changes with the ingress controller

### LINKED ISSUE #208: modify template api definition should trigger update in ingress resource
When creating an Ingress Resource, if I modify the referenced template - e.g. change `use_keyless` to false, I expect all associated API Definitions to be updated.

Currently, this does not happen, making it difficult to update an Ingress resource.

### LINKED ISSUE #217: ingress controller should be able to listen to custom ingress classes
At present, it is only possible for the ingress controller to listen to a specific ingress class `tyk`.

```
kubernetes.io/ingress.class: tyk
```

This should be configurable / overridable based on operator startup configurations.

e.g. 

```
kubernetes.io/ingress.class: nginx
```

or

```
kubernetes.io/ingress.class: tyk-staging
```

### PR DESCRIPTION
**Pull request by gernest**
closes #208 
closes #217

### CODE REVIEW

**Comment by asoorm** [comment]
There is a problem with this implementation. Multiple ingress resources in the same namespace should be able to use the same api definition template. An ingress resource does not own a template.

---

Ingress 1:
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin-ingress1
  annotations:
    kubernetes.io/ingress.class: tyk
    tyk.io/template: myapideftemplate
spec:
  rules:
    - host: httpbin1
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: httpbin
                port:
                  number: 8000
```

Ingress 2:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin-ingress2
  annotations:
    kubernetes.io/ingress.class: tyk
    tyk.io/template: myapideftemplate
spec:
  rules:
    - host: httpbin2
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: httpbin
                port:
                  number: 8000
```

Template:
```
apiVersion: tyk.tyk.io/v1alpha1
kind: ApiDefinition
metadata:
  name: myapideftemplate
  labels:
    template: "true"
spec:
  name: foo
  protocol: http
  use_keyless: true
  active: true
  proxy:
    target_url: http://example.com
    strip_listen_path: true
  version_data:
    default_version: Default
    not_versioned: true
    versions:
      Default:
        name: Default
        paths:
          black_list: []
          ignored: []
          white_list: []
```

Error:

```
{"level":"error","ts":1609257933.2168314,"logger":"controller","msg":"Reconciler error","reconcilerGroup":"networking.k8s.io","reconcilerKind":"Ingress","controller":"ingress","name":"httpbin-ingress2","namespace":"default","error":"Object default/myapideftemplate is already owned by another Ingress controller httpbin-ingress1","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:246\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:197\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90"}
```

File: controllers/ingress_controller.go [hunk 1]
```diff
@@ -56,250 +61,190 @@ type IngressReconciler struct {
...
-	return fmt.Sprintf("%s-%s-%s", nameSpace, name, shortHash(hostName+path))
+func (r *IngressReconciler) buildAPIName(nameSpace, name, hash string) string {
+	return fmt.Sprintf("%s-%s-%s", nameSpace, name, hash)
 }
 
 func shortHash(txt string) string {
 	h := sha256.New()
 	h.Write([]byte(txt))
-	return fmt.Sprintf("%x", h.Sum(nil))[:9]
+	return fmt.Sprintf("%x", h.Sum(nil))[:4]
```
Comment (asoorm) [inline controllers/ingress_controller.go:218]: Why change the API Name / Hash length? What is benefit?

**Review by asoorm — CHANGES_REQUESTED** [review]
There is a problem with this implementation. Multiple ingress resources in the same namespace should be able to use the same api definition template. An ingress resource does not own a template.

---

Ingress 1:
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin-ingress1
  annotations:
    kubernetes.io/ingress.class: tyk
    tyk.io/template: myapideftemplate
spec:
  rules:
    - host: httpbin1
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: httpbin
                port:
                  number: 8000
```

Ingress 2:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin-ingress2
  annotations:
    kubernetes.io/ingress.class: tyk
    tyk.io/template: myapideftemplate
spec:
  rules:
    - host: httpbin2
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: httpbin
                port:
                  number: 8000
```

Template:
```
apiVersion: tyk.tyk.io/v1alpha1
kind: ApiDefinition
metadata:
  name: myapideftemplate
  labels:
    template: "true"
spec:
  name: foo
  protocol: http
  use_keyless: true
  active: true
  proxy:
    target_url: http://example.com
    strip_listen_path: true
  version_data:
    default_version: Default
    not_versioned: true
    versions:
      Default:
        name: Default
        paths:
          black_list: []
          ignored: []
          white_list: []
```

Error:

```
{"level":"error","ts":1609257933.2168314,"logger":"controller","msg":"Reconciler error","reconcilerGroup":"networking.k8s.io","reconcilerKind":"Ingress","controller":"ingress","name":"httpbin-ingress2","namespace":"default","error":"Object default/myapideftemplate is already owned by another Ingress controller httpbin-ingress1","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:246\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:218\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:197\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/ahmet/go/src/github.com/TykTechnologies/tyk-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90"}
```

File: controllers/ingress_controller.go (hunk 1 above)
Comment (gernest) [reply controllers/ingress_controller.go:218]: Nothing much, I just changed to what I originally suggested(first 4 bytes) . It looks much better.

Comment (asoorm) [reply controllers/ingress_controller.go:218]: does this not increase the chance of collision?

Comment (gernest) [reply controllers/ingress_controller.go]: reverted back at 677be00

File: controllers/apidefinition_controller.go [hunk 2]
```diff
@@ -187,32 +197,44 @@ func (r *ApiDefinitionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
...
-		},
+	ls := v1beta1.IngressList{}
+	err := r.List(ctx, &ls,
+		client.InNamespace(ns),
+	)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	for _, v := range ls.Items {
+		if v.GetAnnotations()[ingressTemplateAnnotationKey] == a.Name {
```
Comment (asoorm) [inline controllers/apidefinition_controller.go:216]: This breaks the API. It is not a label, it's now an annotation. And the `ingressTemplateAnnotationKey` is `tyk.io/template: tyk`, not `template: tyk`

File: pkg/environmet/env.go [hunk 3]
```diff
@@ -30,6 +30,13 @@ const (
 
 	// SkipVerify the client will skip tls verification if this is true
 	SkipVerify = "TYK_TLS_INSECURE_SKIP_VERIFY"
+	// IngressClasses is comma separated list of ingress class name that the
+	// operator will listen to. These will added to the default value tyk.
+	//
+	// Example if you set this WATCH_INGRESS_CLASSES="nginx,custom". Then the
+	// operator will start checking tyk,nginx and custom, class name, any event
+	// that has one of the class will reconciled
+	IngressClasses = "WATCH_INGRESS_CLASSES"
```
Comment (asoorm) [inline pkg/environmet/env.go:39]: The ingress controller should only change 1 ingress class.

**Review by asoorm — CHANGES_REQUESTED** [review]

**Comment by asoorm** [comment]
It looks like it is possible to delete a template, despite the fact that an ingress depends on it.


```
k get tykapis
NAME                                 DOMAIN     LISTENPATH   PROXY.TARGETURL                                 ENABLED
default-httpbin-ingress1-e0bf94ae1   httpbin1   /            http://httpbin.default.svc.cluster.local:8000   true
default-httpbin-ingress2-3e8322b04   httpbin2   /            http://httpbin.default.svc.cluster.local:8000   true
myapideftemplate                                             http://example.com                              true
```

```
k delete tykapis myapideftemplate
apidefinition.tyk.tyk.io "myapideftemplate" deleted
```

```
k get tykapis
NAME                                 DOMAIN     LISTENPATH   PROXY.TARGETURL                                 ENABLED
default-httpbin-ingress1-e0bf94ae1   httpbin1   /            http://httpbin.default.svc.cluster.local:8000   true
default-httpbin-ingress2-3e8322b04   httpbin2   /            http://httpbin.default.svc.cluster.local:8000   true
```

```
k get ingress httpbin-ingress1 -o json | jq '.metadata.annotations'
{
  "kubectl.kubernetes.io/last-applied-configuration": "TRUNCATED - NOT IMPORTANT",
  "kubernetes.io/ingress.class": "tyk",
  "tyk.io/template": "myapideftemplate" <-----------------
}
```

File: controllers/ingress_controller.go [hunk 4]
```diff
@@ -254,52 +213,37 @@ func shortHash(txt string) string {
 }
 
 func (r *IngressReconciler) ingressClassEventFilter() predicate.Predicate {
-	isOurIngress := func(annotations map[string]string) bool {
-		if ingressClass, ok := annotations[ingressClassAnnotationKey]; !ok {
+	ls := append([]string{defaultIngressClassAnnotationValue},
+		r.UniversalClient.Environment().IngressClasses...)
```
[the remark this answers is not in the document]
Comment (gernest) [reply controllers/ingress_controller.go]: fixed at 09ac635

File: pkg/environmet/env.go (hunk 3 above)
Comment (gernest) [reply pkg/environmet/env.go]: fixed at 09ac635

File: pkg/environmet/env.go [hunk 5]
```diff
@@ -30,13 +30,9 @@ const (
...
-	// IngressClasses is comma separated list of ingress class name that the
-	// operator will listen to. These will added to the default value tyk.
-	//
-	// Example if you set this WATCH_INGRESS_CLASSES="nginx,custom". Then the
-	// operator will start checking tyk,nginx and custom, class name, any event
-	// that has one of the class will reconciled
-	IngressClasses = "WATCH_INGRESS_CLASSES"
+
+	// IngressClass overides the default class to watch for ingress
+	IngressClass = "WATCH_INGRESS_CLASSES"
```
Comment (asoorm) [inline pkg/environmet/env.go:35]: Nice, just one thing - Env var needs changing:

From
```
	IngressClass = "WATCH_INGRESS_CLASSES"
```

To

```
	IngressClass = "WATCH_INGRESS_CLASS"
```

**Review by asoorm — CHANGES_REQUESTED** [review]

**Review by asoorm — APPROVED** [review]

**Pull request merged**